src/Blocks/BlockButton.h

```diff
@@ -133,12 +123,119 @@ class cBlockButtonHandler :
+	static void OnArrowHit(cWorld & a_World, const Vector3i a_Position, const eBlockFace a_HitFace)
+	{
+		BLOCKTYPE Type;
+		NIBBLETYPE Meta;
+		const auto Pos = AddFaceDirection(a_Position, a_HitFace);
+
+		if (
+			!a_World.GetBlockTypeMeta(Pos, Type, Meta) ||
+			IsButtonOn(Meta) ||
+			!IsButtonPressedByArrow(a_World, Pos, Type, Meta)
+		)
+		{
+			// Bail if we're not specifically a wooden button, or it's already on
+			// or if the arrow didn't intersect. It is very important that nothing is
+			// done if the button is depressed, since the release task will already be queued
+			return;
+		}
+
+		a_World.SetBlockMeta(Pos, Meta | 0x08, false);
+		a_World.WakeUpSimulators(Pos);
+
+		// sound name is ok to be wood, because only wood gets triggered by arrow
+		a_World.GetBroadcastManager().BroadcastSoundEffect("block.wood_button.click_on", Pos, 0.5f, 0.6f);
+
+		// Queue a button reset
+		QueueButtonRelease(a_World, Pos, Type);
+	}
+
+private:
+
+	/** Schedules a recurring event at appropriate intervals to release a button at a given position.
+	The given block type is checked when the task is executed to ensure the position still contains a button. */
+	static void QueueButtonRelease(cWorld & a_ButtonWorld, const Vector3i a_Position, const BLOCKTYPE a_BlockType)
+	{
+		const auto TickDelay = (a_BlockType == E_BLOCK_STONE_BUTTON) ? 20 : 30;
+		a_ButtonWorld.ScheduleTask(
+			TickDelay,
+			[a_Position, a_BlockType](cWorld & a_World)
+			{
+				BLOCKTYPE Type;
+				NIBBLETYPE Meta;
+
+				if (
+					!a_World.GetBlockTypeMeta(a_Position, Type, Meta) ||
+					(Type != a_BlockType) || !IsButtonOn(Meta)
+				)
+				{
+					// Total failure or block changed, bail
+					return;
+				}
+
+				if (IsButtonPressedByArrow(a_World, a_Position, Type, Meta))
+				{
+					// Try again in a little while
+					QueueButtonRelease(a_World, a_Position, a_BlockType);
+					return;
+				}
+
+				// Block hasn't change in the meantime; release it
+				const auto SoundToPlayOnRelease = (Type == E_BLOCK_STONE_BUTTON) ? "block.stone_button.click_off" : "block.wood_button.click_off";
+
+				a_World.SetBlockMeta(a_Position, Meta & 0x07, false);
+				a_World.WakeUpSimulators(a_Position);
+				a_World.BroadcastSoundEffect(SoundToPlayOnRelease, a_Position, 0.5f, 0.5f);
+			}
+		);
+	}
+
+	/** Returns true if an arrow was found in the wooden button */
+	static bool IsButtonPressedByArrow(cWorld & a_World, const Vector3i a_ButtonPosition, const BLOCKTYPE a_BlockType, const NIBBLETYPE a_Meta)
+	{
+		if (a_BlockType != E_BLOCK_WOODEN_BUTTON)
+		{
+			return false;
+		}
+
+		const auto FaceOffset = GetButtonOffsetOnBlock(a_Meta);
+		const bool FoundArrow = !a_World.ForEachEntityInBox(
+			cBoundingBox(FaceOffset + a_ButtonPosition, 0.2, 0.2),
+			[](cEntity & a_Entity)
+			{
+				return a_Entity.IsArrow();
+			}
+		);
+
+		return FoundArrow;
+	}
+
+	static Vector3d GetButtonOffsetOnBlock(NIBBLETYPE a_Meta)
```

Comment:
This requires a doxycomment on what the meta is (why the YM case returns the *top* ?)